components/raftstore-v2/src/operation/query/lease.rs

```diff
@@ -0,0 +1,352 @@
+        read.addition_request = request.map(Box::new);
+        self.pending_reads_mut().push_back(read, is_leader);
+        debug!(
+            self.logger,
+            "request to get a read index";
+            "request_id" => ?id,
+            "is_leader" => is_leader,
+        );
+
+        true
+    }
+
+    fn leader_read_index<T: Transport>(
+        &mut self,
+        poll_ctx: &mut StoreContext<EK, ER, T>,
+        mut req: RaftCmdRequest,
+        ch: QueryResChannel,
+        now: Timespec,
+    ) -> bool {
+        let lease_state = self.inspect_lease();
+        if can_amend_read::<Peer<EK, ER>, QueryResChannel>(
+            self.pending_reads().back(),
+            &req,
+            lease_state,
+            poll_ctx.cfg.raft_store_max_leader_lease(),
+            now,
+        ) {
+            // Must use the commit index of `PeerStorage` instead of the commit index
+            // in raft-rs which may be greater than the former one.
+            // For more details, see the annotations above `on_leader_commit_idx_changed`.
+            let commit_index = self.storage().entry_storage().commit_index();
+            if let Some(read) = self.pending_reads_mut().back_mut() {
+                // A read request proposed in the current lease is found; combine the new
+                // read request to that previous one, so that no proposing needed.
+                read.push_command(req, ch, commit_index);
+                return false;
+            }
+        }
+
+        if !self.propose_read_index(poll_ctx, req, self.is_leader(), ch, now) {
+            return false;
+        }
+
+        // TimeoutNow has been sent out, so we need to propose explicitly to
+        // update leader lease.
+        // TODO:add following when propose is done
+        // if self.leader_lease.is_suspect() {
+        // let req = RaftCmdRequest::default();
+        // if let Ok(Either::Left(index)) = self.propose_normal(poll_ctx, req) {
+        // let (callback, _) = CmdResChannel::pair();
+        // let p = Proposal {
+        // is_conf_change: false,
+        // index,
+        // term: self.term(),
+        // cb: callback,
+        // propose_time: Some(now),
+        // must_pass_epoch_check: false,
+        // };
+        //
+        // self.post_propose(poll_ctx, p);
+        // }
+        // }
+        true
+    }
+
+    // Returns a boolean to indicate whether the `read` is proposed or not.
+    // For these cases it won't be proposed:
+    // 1. The region is in merging or splitting;
+    // 2. The message is stale and dropped by the Raft group internally;
+    // 3. There is already a read request proposed in the current lease;
+    pub fn read_index<T: Transport>(
+        &mut self,
+        poll_ctx: &mut StoreContext<EK, ER, T>,
+        mut req: RaftCmdRequest,
+        ch: QueryResChannel,
+    ) -> bool {
+        // TODO: add pre_read_index to handle splitting or merging
+        let now = monotonic_raw_now();
```

Comment:
Better call on demand.

Reply:
This is to save one system call, as you suggested before. Original version is put this line to the following two read_index fns.